ppy/osu-queue-score-statistics
# Prepare infrastructure for implementing hush-hush medals

- [x] Depends on https://github.com/ppy/osu-queue-score-statistics/pull/195 for mergeability

RFC. PRing so that the thing I've spent multiple days on now can be looked at closely and accurately without me typing up hand-wavey essays. Will want input from @peppy and @ThePooN.

## How this is supposed to work

The premise is that this project is to remain public, but will consume a private DLL where the hush-hush medals are implemented.

In principle the setup I'm about to describe requires no new nuget packages, but will require separate docker images. The loose sketch of how I imagine the private part is going to look is available for preview at https://github.com/bdach/osu-hush-hush-medals-poc. (The project is intended to be public, it contains no actual code that needs to be kept secret. It's just a PoC.)

This works as follows:

1. The private project consumes the public project at a _source level_ as a submodule.
2. The private project references the public project at a _.NET level_, which means it can consume all of its public API.
3. The public project remains the primary runnable one, but it can load private DLLs with private processors and medal awarders via the `EXTERNAL_PROCESSOR_PATH` envvar.

    I know what you're probably thinking: this means that there is a circular dependency now. Please read on despite this.
4. The private project contains a Dockerfile. That Dockerfile builds an image from the private project, and thus transitively builds the public project too as the private one depends on it. Thus the build output contains both the private and public DLLs, as output by _a single compilation_.

    This is important because it is _transparent_ (as in, you can deal with deployment of only the private version) and _fail-safe_. Yes the proper way to do this would be to move `IProcessor`, `IMedalAwarder`, etc. out to a separate "models" nuget and have the public and private projects reference that, and I initially did this. But that makes it possible for a _binary incompatibility_ of the shared models subproject if you forget to bump one or the other at either end, which then will most likely die at runtime because we're still going to be wanting to load the private DLL. If the two are built together via the transitive dependency, and deployed together, this is _statically_ avoided. If anything breaks, the project just won't build.
5. The private docker image will be the one that will be deployed. Its entrypoint is the public DLL, with the private DLL loaded in as pointed by the envvar.

As part of review of this, please scrutinise the following in the PoC too:
- [`MyFirstMedalAwarder`](https://github.com/bdach/osu-hush-hush-medals-poc/blob/master/osu.Server.Queues.ScoreStatisticsProcessor.HushHushMedals/MyFirstMedalAwarder.cs)
- [`MyFirstMedalAwarderTest`](https://github.com/bdach/osu-hush-hush-medals-poc/blob/master/osu.Server.Queues.ScoreStatisticsProcessor.HushHushMedals.Tests/MyFirstMedalAwarderTest.cs)
- [`Dockerfile`](https://github.com/bdach/osu-hush-hush-medals-poc/blob/master/Dockerfile)

## Testing

The PoC above can be tested full-stack with a local osu-web checkout setup.

1. Clone https://github.com/bdach/osu-hush-hush-medals-poc.
2. Ensure that the LIO endpoint is pointed at localhost (in the _submodule_ of the poc) and that `SHARED_INTEROP_SECRET` is set web-side.
3. Change [`MyFirstMedalAwarder`](https://github.com/bdach/osu-hush-hush-medals-poc/blob/master/osu.Server.Queues.ScoreStatisticsProcessor.HushHushMedals/MyFirstMedalAwarder.cs#L16) to point at a valid medal ID in your db. (It'll attempt to grant the medal after every completed play until it succeeds. If it succeeds and you wanna try again without rebuilding, just truncate `osu_user_achievements`.)
4. Build the image:
    ```sh
    $ docker build . -t osu-queue-score-statistics-private
    ```
5. Run the image:
    ```sh
    $ docker run --rm --net=host -e 'SHARED_INTEROP_SECRET=${your-secret}' osu-queue-score-statistics-private queue watch
    ```
    (`--net=host` is used so that the container can see the ports that web docker compose forwarded out. I'm not sure if there's a better way to do this.)
6. Launch game and finish any beatmap. You should see the medal you chose in point (3) be awarded.
